### PR TITLE
Fix targeted job quota calculation for given up tasks

### LINKED ISSUE #1547: Fix targeted jobs quota occupation for given up tasks
### Describe the bug
For fixed targeted jobs, if the tasks have been given up and reached their terminal state, they might still occupy the qupta and block the job to have progress for other unassigned tasks.

### To Reproduce
Start an instance with more than 40 partitions (e.g. 100 or something). Then start a targeted job which all of its tasks needs to be assigned to the only available instance. Tasks need to be failed (go to TASK_ERROR state right away) and have a maximum number of attempts of 2 for example. You would notice that the job would not finish.

### Expected behavior
The given up tasks should not go to the fixed targeted task assignment for scheduling and should not occupy the quota.

### PR DESCRIPTION
**Pull request by alirezazamani**
### Issues

- [X] My PR addresses the following Helix issues and references them in the PR description:
Fixes #1547 

### Description

- [x] Here are some details about my PR, including screenshots of any UI changes:
In this PR, the tasks that should not be retried again, will not occupy any quota. The would avoid the jobs
being blocked because of quota usage of given up tasks.

### Tests

- [x] The following tests are written for this issue:
TestTaskQuotaCalculations.testTaskErrorMaxRetriesQuotaRelease

- [x] The following is the result of the "mvn test" command on the appropriate module:
Helix-core:
```
[ERROR] org.apache.helix.tools.TestClusterStateVerifier.beforeMethod(org.apache.helix.tools.TestClusterStateVerifier)
[ERROR]   Run 1: TestClusterStateVerifier.beforeMethod:57 » Helix cluster null is not setup yet
[INFO]   Run 2: PASS
[INFO]
[INFO]
[ERROR] Tests run: 1253, Failures: 1, Errors: 0, Skipped: 3
[INFO]
[INFO] ------------------------------------------------------------------------
[INFO] BUILD FAILURE
[INFO] ------------------------------------------------------------------------
[INFO] Total time:  01:29 h
[INFO] Finished at: 2020-11-22T19:02:28-08:00
[INFO] ------------------------------------------------------------------------
```
The failed test is unrelated to this change and is failing even without this PR.

Helix-rest:
```
[INFO] Tests run: 171, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 246.03 s - in TestSuite
[INFO]
[INFO] Results:
[INFO]
[INFO] Tests run: 171, Failures: 0, Errors: 0, Skipped: 0
[INFO]
[INFO] ------------------------------------------------------------------------
[INFO] BUILD SUCCESS
[INFO] ------------------------------------------------------------------------
[INFO] Total time:  04:10 min
[INFO] Finished at: 2020-11-22T22:21:02-08:00
[INFO] ------------------------------------------------------------------------
```

### Commits

- My commits all reference appropriate Apache Helix GitHub issues in their subject lines. In addition, my commits follow the guidelines from "[How to write a good git commit message](http://chris.beams.io/posts/git-commit/)":
  1. Subject is separated from body by a blank line
  1. Subject is limited to 50 characters (not including Jira issue reference)
  1. Subject does not end with a period
  1. Subject uses the imperative mood ("add", not "adding")
  1. Body wraps at 72 characters
  1. Body explains "what" and "why", not "how"

### Code Quality

- My diff has been formatted using helix-style.xml 
(helix-style-intellij.xml if IntelliJ IDE is used)

### CODE REVIEW

File: helix-core/src/main/java/org/apache/helix/task/AbstractTaskDispatcher.java [hunk 1]
```diff
@@ -838,7 +839,7 @@ private static void addCompletedTasks(Set<Integer> set, JobContext ctx, Iterable
    */
   private boolean isTaskNotInTerminalState(TaskPartitionState state) {
```
Comment (jiajunwang) [inline helix-core/src/main/java/org/apache/helix/task/AbstractTaskDispatcher.java:840]: This change also impacts filterTasks(). The logic looks to be valid even after this change.
The concerning part is that there are some duplicate checks here and there. For example,

      // Allow tasks eligible for scheduling
      if (state == null || state == TaskPartitionState.STOPPED
          || state == TaskPartitionState.TIMED_OUT || state == TaskPartitionState.TASK_ERROR
          || state == TaskPartitionState.DROPPED) {
        filteredTasks.add(partitionNumber);
      }
      // Allow tasks whose assigned instances are no longer live for rescheduling
      if (isTaskNotInTerminalState(state)) {
        ...
      }

This section checks for different state combinations and then react differently. First of all, it is hard for the reviewers to follow the logic. Secondly, it is very easy for the coder to introduce bugs since the combinations are separately done in different private methods.
Is it possible that we conclude all the states check within 3 or 4 methods? Or we need to clean up the state. The current design is not reviewable.

Comment (alirezazamani) [reply helix-core/src/main/java/org/apache/helix/task/AbstractTaskDispatcher.java:840]: We need to note that this no design. This is a bug which is introduced before and is impacting production result. I tried differently but I believe this is the best way.

File: helix-core/src/test/java/org/apache/helix/integration/task/TestTaskErrorMaxRetriesQuotaRelease.java [hunk 2]
```diff
@@ -0,0 +1,63 @@
...
+import org.apache.helix.task.JobQueue;
+import org.apache.helix.task.TaskState;
+import org.apache.helix.task.TaskUtil;
+import org.testng.annotations.AfterClass;
+import org.testng.annotations.BeforeClass;
+import org.testng.annotations.Test;
+
+import com.google.common.collect.ImmutableMap;
+
+public class TestTaskErrorMaxRetriesQuotaRelease extends TaskTestBase {
```
[the remark this answers is not in the document]
Comment (alirezazamani) [reply helix-core/src/test/java/org/apache/helix/integration/task/TestTaskErrorMaxRetriesQuotaRelease.java:33]: I don't think it is possible. It specifically needs a new setup. One instance with a lot of partitions. Adding it to new tests would make other test logic change.

File: helix-core/src/test/java/org/apache/helix/integration/task/TestTaskErrorMaxRetriesQuotaRelease.java [hunk 3]
```diff
@@ -0,0 +1,63 @@
...
+  @BeforeClass
+  public void beforeClass() throws Exception {
+    _numNodes = 1;
+    _numPartitions = 100;
+    super.beforeClass();
+  }
+
+  @AfterClass
+  public void afterClass() throws Exception {
+    super.afterClass();
```
[the remark this answers is not in the document]
Comment (alirezazamani) [reply helix-core/src/test/java/org/apache/helix/integration/task/TestTaskErrorMaxRetriesQuotaRelease.java:44]: Ok, I will remove it.

File: helix-core/src/main/java/org/apache/helix/task/AbstractTaskDispatcher.java (hunk 1 above)
Comment (alirezazamani) [reply helix-core/src/main/java/org/apache/helix/task/AbstractTaskDispatcher.java:840]: Also here we are trying to minimize the usage of TaskPartitionState.... in this function and move the logic to other functions. If not, the code will be ugly and hard to follow. By just using given up, we mean this task is not reliable by the controller.

Comment (alirezazamani) [reply helix-core/src/main/java/org/apache/helix/task/AbstractTaskDispatcher.java:840]: One more thing to add. I think the designer of this method (I was not involved in designing this method BTW) used the variable `partitionsToRetryOnLiveInstanceChangeForTargetedJob` to give some hint to the reader about what are these tasks and why we are adding them back. Hence, I think keeping it as it is still the best solution.

File: helix-core/src/test/java/org/apache/helix/integration/task/TestTaskErrorMaxRetriesQuotaRelease.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (alirezazamani) [reply helix-core/src/test/java/org/apache/helix/integration/task/TestTaskErrorMaxRetriesQuotaRelease.java:33]: Basically, I use this theory in my tests. Write a test that fails without this PR. Add the PR and make sure the test will be passed now. The setup to have a failing test is not available in other tests. That is why I need to have a new setup (scenario) for this PR.

[the remark this answers is not in the document]
Comment (alirezazamani) [reply helix-core/src/test/java/org/apache/helix/integration/task/TestTaskErrorMaxRetriesQuotaRelease.java]: Move to another exiting classes.

**Review by jiajunwang — APPROVED** [review]
LGTM!

**Comment by alirezazamani** [comment]
This PR is ready to be merged. 

Final commit message:
Fix targeted job quota calculation for given up tasks

In this commit, the tasks that should not be retried again, will not occupy 
any quota. The would avoid the jobs being blocked because of quota 
usage of given up tasks.
